bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+//! Checks for function parameters that take a mutable reference to a
+//! re-borrwable type.
+//!
+//! # Motivation
+//!
+//! Mutable references to re-borrwable types can almost always be readily
+//! converted back to an owned instance of itself, albeit with an appropriately
+//! shorter lifetime.
+//!
+//! The only time this isn't true is when the function returns referenced data
+//! that is bound to the mutable reference of the re-borrowable type.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(&mut commands);
+//! }
+//!
+//! fn helper_function(commands: &mut Commands) {
+//!   // ...
+//! }
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(commands.reborrow());
+//! }
+//!
+//! fn helper_function(mut commands: Commands) {
+//!   // ...
+//! }
+//! ```
+
+use std::ops::ControlFlow;
+
+use crate::declare_bevy_lint;
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::match_type};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, Mutability};
+use rustc_lint::{LateContext, LateLintPass, Lint};
+use rustc_middle::{
+    ty::Interner,
+    ty::{Ty, TyKind, TypeVisitable, TypeVisitor},
+};
+use rustc_session::declare_lint_pass;
+use rustc_span::{def_id::LocalDefId, symbol::Ident, Span};
+
+declare_bevy_lint! {
+    pub BORROW_OF_COMMANDS,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Commands` or `EntityCommands` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_RESOURCE,
+    PEDANTIC,
+    "parameter takes a mutable reference to `ResMut` or `NonSendMut` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_QUERY,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Query` instead of a re-borrowed instance",
+}
+
+declare_lint_pass! {
+    BorrowOfReborrowable => [BORROW_OF_COMMANDS.lint, BORROW_OF_RESOURCE.lint, BORROW_OF_QUERY.lint]
+}
+
+impl<'tcx> LateLintPass<'tcx> for BorrowOfReborrowable {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // We are already inside of the function item,
+        // so we can use `instantiate_identity` to discharge the binder
+        let fn_sig = cx.tcx.fn_sig(def_id).instantiate_identity();
+
+        let arg_names = cx.tcx.fn_arg_names(def_id);
+
+        fn_sig.inputs().map_bound(|args| {
```

Comment:
`map_bound()` is used to transform the type contained by the binder, which in this case is `&[Ty]`. This closure doesn't actually return anything, though, so `map_bound()` just ends up doing extra work. Looking at the source you can see:

```rust
pub fn map_bound<F, U: TypeVisitable<I>>(self, f: F) -> Binder<I, U>
where
    F: FnOnce(T) -> U,
{
    // Break apart the binder into its value and bound variables.
    let Binder { value, bound_vars } = self;
    // Transform the value, ignoring the bound vars.
    let value = f(value);
    // Debug checks to make sure we aren't committing type crimes.
    if cfg!(debug_assertions) {
        let mut validator = ValidateBoundVars::new(bound_vars);
        value.visit_with(&mut validator);
    }
    // Returning the new, mapped binder.
    Binder { value, bound_vars }
}
```

In this code, the closure completely ignores `bound_vars` until its returned value gets checked. Because the closure returns nothing, we have no purpose messing around with `bound_vars` in the first place.

So in conclusion: since we do not need `bound_vars`, can we switch this to:

```suggestion
        let args = fn_sig.inputs().skip_binder();
```

and inline the `for` loop directly into the main function?

Reply:
It's funny you say that. I originally used `skip_binder`, but then the [docs](https://doc.rust-lang.org/stable/nightly-rustc/rustc_type_ir/binder/struct.Binder.html#method.skip_binder) scared me into using something that sounded safer (i.e. `map_bound`).

I'll change it back!